cmd/sfncli/ecs.go

```diff
@@ -1,133 +1,66 @@
-		if err != nil {
-			break
-		}
-
-		components := strings.Split(line, ":")
-		if len(components) == 3 {
-			return strings.TrimRight(path.Base(components[2]), "\n"), nil
-		}
-	}
-
-	return "", fmt.Errorf("Failed to find Docker ID in /proc/self/group")
-}
-
-type ECSAgentMetadata struct {
-	Cluster string `json:"Cluster"`
-}
-
-type ECSAgentTaskMetadata struct {
-	Tasks []struct {
-		ARN        string `json:"Arn"`
-		Containers []struct {
-			DockerID string `json:"DockerId"`
-		} `json:"Containers"`
-	} `json:"Tasks"`
-}
-
-// https://github.com/aws/amazon-ecs-agent/issues/258
-// https://github.com/aws/amazon-ecs-agent/pull/709
-func ecsAgentTaskMetadata() (ECSAgentTaskMetadata, error) {
-	response, err := http.Get("http://172.17.0.1:51678/v1/tasks")
-	if err != nil {
-		return ECSAgentTaskMetadata{}, err
-	}
-	defer response.Body.Close()
-	body, err := ioutil.ReadAll(response.Body)
-	if err != nil {
-		return ECSAgentTaskMetadata{}, err
-	}
-
-	metadata := ECSAgentTaskMetadata{}
-	err = json.Unmarshal(body, &metadata)
-	if err != nil {
-		return ECSAgentTaskMetadata{}, err
-	}
-
-	return metadata, nil
-}
-
-func ecsAgentMetadata() (ECSAgentMetadata, error) {
-	response, err := http.Get("http://172.17.0.1:51678/v1/metadata")
-	if err != nil {
-		return ECSAgentMetadata{}, err
-	}
-	defer response.Body.Close()
-	body, err := ioutil.ReadAll(response.Body)
-	if err != nil {
-		return ECSAgentMetadata{}, err
-	}
-
-	metadata := ECSAgentMetadata{}
-	err = json.Unmarshal(body, &metadata)
-	if err != nil {
-		return ECSAgentMetadata{}, err
-	}
-
-	return metadata, nil
+// ecsContainerMetadata is a subset of fields in the container metadata file. Doc for reference:
+// https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-metadata.html#metadata-file-format
+type ecsContainerMetadata struct {
+	TaskARN            string
+	MetadataFileStatus string
 }
 
+// expandECSTaskARN uses ECS Container Metadata to magically populate the TaskARN required to
+// register with AWS Step Functions. Doc for reference:
+// https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-metadata.html
 func expandECSTaskARN(s string) (string, error) {
 	if !strings.Contains(s, magicECSTaskARN) {
 		return s, nil
 	}
 
-	dockerID, err := getDockerID()
-	if err != nil {
-		return "", err
+	// this is an env var the AWS ECS agent sets for us
+	filePath, ok := os.LookupEnv("ECS_CONTAINER_METADATA_FILE")
+	if !ok {
+		return "", fmt.Errorf("ECS_CONTAINER_METADATA_FILE is not set")
 	}
 
-	agentMetadata, err := ecsAgentMetadata()
-	if err != nil {
-		return "", err
-	}
-
-	if agentMetadata.Cluster == "" {
-		return "", fmt.Errorf("Could not find ECS cluster for docker container '%s'", dockerID)
-	}
-
-	agentTaskMetadata, err := ecsAgentTaskMetadata()
-	if err != nil {
-		return "", err
+	// wait for the file to exist
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		<-ticker.C
+		if _, err := os.Stat(filePath); err == nil {
+			break
+		}
 	}
```

Comment:
looks like this waits forever, which doesn't seem ideal...

Reply:
Hmmm I'm not too sure about that - here's my local flow https://gist.github.com/Sayan-/ffdaac6ec0d9f87cb218da50150362b2